<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BayesSpace • BayesSpace</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BayesSpace">
<meta property="og:description" content="BayesSpace">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BayesSpace</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/BayesSpace.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Examples
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/maynard_DLPFC.html">DLPFC (Maynard et al., 2020)</a>
    </li>
    <li>
      <a href="../articles/ji_SCC.html">SCC (Ji et al., 2020)</a>
    </li>
    <li>
      <a href="../articles/thrane_melanoma.html">Melanoma (Thrane et al., 2018)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/edward130603/BayesSpace">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="BayesSpace_files/header-attrs-2.3/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>BayesSpace</h1>
                        <h4 class="author">Edward Zhao, Matt Stone, Xing Ren, and Raphael Gottardo</h4>
            
            <h4 class="date">2020-10-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/edward130603/BayesSpace/blob/master/vignettes/BayesSpace.Rmd"><code>vignettes/BayesSpace.Rmd</code></a></small>
      <div class="hidden name"><code>BayesSpace.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">BayesSpace</span>)
</pre></div>
<div id="preparing-your-experiment-for-bayesspace" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-your-experiment-for-bayesspace" class="anchor"></a>Preparing your experiment for BayesSpace</h2>
<div id="loading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h3>
<p>BayesSpace supports three ways of loading a <code>SingleCellExperiment</code> for analysis.</p>
<p>Visium datasets processed with <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger">Space Ranger</a> can be loaded directly via the <code><a href="../reference/readVisium.html">readVisium()</a></code> function. This function takes only the path to the Space Ranger output directory (containing the <code>spatial/</code> and <code>filtered_feature_bc_matrix/</code> subdirectories) and returns a <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readVisium.html">readVisium</a></span>(<span class="st">"path/to/spaceranger/outs/"</span>)
</pre></div>
<p>Second, all datasets analyzed for the BayesSpace manuscript are readily accessible via the <code><a href="../reference/getRDS.html">getRDS()</a></code> function. This function takes two arguments - the name of the dataset, and the name of the sample in the dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRDS.html">getRDS</a></span>(dataset=<span class="st">"2018_thrane_melanoma"</span>, sample=<span class="st">"ST_mel1_rep2"</span>)
</pre></div>
<p>Finally, <code>SingleCellExperiment</code> objects can be constructed manually from a counts matrix and tables of row and column data. BayesSpace only requires that spot array coordinates be provided as columns named <code>row</code> and <code>col</code> in <code>colData</code>. (Note that enhancement of Visium datasets additionally requires the pixel coordinates of each spot in the tissue image, but in this case the dataset should be loaded with <code><a href="../reference/readVisium.html">readVisium()</a></code>, which loads these data automatically.)</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="kw">library</span>(Matrix)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>rowData &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">"path/to/rowData.csv"</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>colData &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">"path/to/colData.csv"</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>, <span class="dt">row.names=</span><span class="dv">1</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a>counts &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">"path/to/counts.csv.gz"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a>                   <span class="dt">row.names=</span><span class="dv">1</span>, <span class="dt">check.names=</span>F, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)<span class="er">)</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a>sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays=</span><span class="kw">list</span>(<span class="dt">counts=</span><span class="kw">as</span>(counts, <span class="st">"dgCMatrix"</span>)),</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a>                            <span class="dt">rowData=</span>rowData,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a>                            <span class="dt">colData=</span>colData)</span></code></pre></div>
<p>We’ll continue with the melanoma sample from the 2018 Spatial Transcriptomics paper for the remaining examples in this vignette.</p>
</div>
<div id="pre-processing-data" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-processing-data" class="anchor"></a>Pre-processing data</h3>
<p>BayesSpace requires minimal data pre-processing, but we provide a helper function to automate it.</p>
<p><code><a href="../reference/spatialPreprocess.html">spatialPreprocess()</a></code> log-normalizes the count matrix and performs PCA on the top <code>n.HVGs</code> highly variable genes, keeping the top <code>n.PCs</code> principal components. Additionally, the spatial sequencing platform is added as metadata in the <code>SingleCellExperiment</code> for downstream analyses. If you do not wish to rerun PCA, running <code><a href="../reference/spatialPreprocess.html">spatialPreprocess()</a></code> with the flag <code>skip.PCA=TRUE</code> will only add the metadata BayesSpace requires.</p>
<p>Here, we omit log-normalization as all datasets available through <code><a href="../reference/getRDS.html">getRDS()</a></code> already include log-normalized counts.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">102</span>)
<span class="kw">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialPreprocess.html">spatialPreprocess</a></span>(<span class="kw">melanoma</span>, platform=<span class="st">"ST"</span>, 
                              n.PCs=<span class="fl">7</span>, n.HVGs=<span class="fl">2000</span>, log.normalize=<span class="fl">FALSE</span>)
</pre></div>
</div>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h2>
<div id="selecting-the-number-of-clusters" class="section level3">
<h3 class="hasAnchor">
<a href="#selecting-the-number-of-clusters" class="anchor"></a>Selecting the number of clusters</h3>
<p>We can use the <code><a href="../reference/qTune.html">qTune()</a></code> and <code><a href="../reference/qTune.html">qPlot()</a></code> functions to help choose <code>q</code>, the number of clusters to use in our analysis.</p>
<ul>
<li>
<code><a href="../reference/qTune.html">qTune()</a></code> runs the BayesSpace clustering algorithm for multiple specified values of <code>q</code> (by default, 3 through 7) and computes their average pseudo-log-likelihood. It accepts any arguments to <code><a href="../reference/spatialCluster.html">spatialCluster()</a></code>.</li>
<li>
<code><a href="../reference/qTune.html">qPlot()</a></code> plots the pseudo-log-likelihood as a function of <code>q</code>; we suggest choosing a <code>q</code> around the elbow of this plot.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qTune.html">qTune</a></span>(<span class="kw">melanoma</span>, qs=<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">2</span>, <span class="fl">10</span>), platform=<span class="st">"ST"</span>, d=<span class="fl">7</span>)
<span class="fu"><a href="../reference/qTune.html">qPlot</a></span>(<span class="kw">melanoma</span>)
</pre></div>
<p><img src="figures/BayesSpace-tuning_q-1.png" width="100%"></p>
</div>
<div id="clustering-with-bayesspace" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering-with-bayesspace" class="anchor"></a>Clustering with BayesSpace</h3>
<p>The <code><a href="../reference/spatialCluster.html">spatialCluster()</a></code> function clusters the spots, and adds the predicted cluster labels to the <code>SingleCellExperiment</code>. Typically, as we did for the analyses in the paper, we suggest running with at least 10,000 iterations (<code>nrep=10000</code>), but we use 1,000 iteration in this demonstration for the sake of runtime. (Note that a random seed must be set in order for the results to be reproducible.)</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">149</span>)
<span class="kw">melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialCluster.html">spatialCluster</a></span>(<span class="kw">melanoma</span>, q=<span class="fl">4</span>, platform=<span class="st">"ST"</span>, d=<span class="fl">7</span>,
                           init.method=<span class="st">"mclust"</span>, model=<span class="st">"t"</span>, gamma=<span class="fl">2</span>,
                           nrep=<span class="fl">1000</span>, burn.in=<span class="fl">100</span>,
                           save.chain=<span class="fl">TRUE</span>)
</pre></div>
<p>Both the mclust initialization (<code>cluster.init</code>) and the BayesSpace cluster assignments (<code>spatial.cluster</code>) are now available in the SingleCellExperiment’s <code>colData</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">colData</span>(<span class="kw">melanoma</span>))
<span class="co">#&gt; DataFrame with 6 rows and 5 columns</span>
<span class="co">#&gt;            row       col sizeFactor cluster.init spatial.cluster</span>
<span class="co">#&gt;      &lt;integer&gt; &lt;integer&gt;  &lt;numeric&gt;    &lt;numeric&gt;       &lt;numeric&gt;</span>
<span class="co">#&gt; 7x15         7        15   0.795588            1               1</span>
<span class="co">#&gt; 7x16         7        16   0.307304            1               1</span>
<span class="co">#&gt; 7x17         7        17   0.331247            2               2</span>
<span class="co">#&gt; 7x18         7        18   0.420747            3               2</span>
<span class="co">#&gt; 8x13         8        13   0.255453            1               1</span>
<span class="co">#&gt; 8x14         8        14   1.473439            1               1</span>
</pre></div>
</div>
<div id="visualizing-spatial-clusters" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-spatial-clusters" class="anchor"></a>Visualizing spatial clusters</h3>
<p>We can plot the cluster assignments over the spatial locations of the spots with <code><a href="../reference/clusterPlot.html">clusterPlot()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="../reference/clusterPlot.html">clusterPlot</a></span>(<span class="kw">melanoma</span>)
</pre></div>
<p><img src="figures/BayesSpace-cluster.plot-1.png" width="100%"></p>
<p>As <code><a href="../reference/clusterPlot.html">clusterPlot()</a></code> returns a <code>ggplot</code> object, it can be customized by composing with familiar <code>ggplot2</code> functions. Additionally, the argument <code>palette</code> sets the colors used for each cluster, and <code><a href="../reference/clusterPlot.html">clusterPlot()</a></code> takes additional arguments to <code><a href="https://ggplot2.tidyverse.org/reference/geom_polygon.html">geom_polygon()</a></code> such as <code>size</code> or <code>color</code> to control the aesthetics of the spot borders.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="../reference/clusterPlot.html">clusterPlot</a></span>(<span class="kw">melanoma</span>, palette=<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"purple"</span>, <span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"yellow"</span>), color=<span class="st">"black"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Column"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"Row"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(fill=<span class="st">"BayesSpace\ncluster"</span>, title=<span class="st">"Spatial clustering of ST_mel1_rep2"</span>)
</pre></div>
<p><img src="figures/BayesSpace-cluster.plot.customize-1.png" width="100%"></p>
</div>
</div>
<div id="enhanced-resolution" class="section level2">
<h2 class="hasAnchor">
<a href="#enhanced-resolution" class="anchor"></a>Enhanced resolution</h2>
<div id="clustering-at-enhanced-resolution" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering-at-enhanced-resolution" class="anchor"></a>Clustering at enhanced resolution</h3>
<p>The <code><a href="../reference/spatialEnhance.html">spatialEnhance()</a></code> function will enhance the resolution of the principal components, and add these PCs as well as predicted cluster labels at subspot resolution to a new <code>SingleCellExperiment</code>. As with our demonstration of <code><a href="../reference/spatialCluster.html">spatialCluster()</a></code> above, we are using fewer iterations for the purpose of this example (<code>nrep=1000</code>) than we recommend in practice (<code>nrep=100000</code> or greater).</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">melanoma.enhanced</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatialEnhance.html">spatialEnhance</a></span>(<span class="kw">melanoma</span>, q=<span class="fl">4</span>, platform=<span class="st">"ST"</span>, d=<span class="fl">7</span>,
                                    model=<span class="st">"t"</span>, gamma=<span class="fl">2</span>,
                                    jitter_prior=<span class="fl">0.3</span>, jitter_scale=<span class="fl">3.5</span>,
                                    nrep=<span class="fl">1000</span>, burn.in=<span class="fl">100</span>,
                                    save.chain=<span class="fl">TRUE</span>)
</pre></div>
<p>The enhanced <code>SingleCellExperiment</code> includes an index to the parent spot in the original <code>sce</code> (<code>spot.idx</code>), along with an index to the subspot. It adds the offsets to the original spot coordinates, and provides the enhanced cluster label (<code>spatial.cluster</code>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">colData</span>(<span class="kw">melanoma.enhanced</span>))
<span class="co">#&gt; DataFrame with 6 rows and 9 columns</span>
<span class="co">#&gt;              spot.idx subspot.idx  spot.row  spot.col       row       col</span>
<span class="co">#&gt;             &lt;numeric&gt;   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="co">#&gt; subspot_1.1         1           1         7        15   7.33333   15.3333</span>
<span class="co">#&gt; subspot_2.1         2           1         7        16   7.33333   16.3333</span>
<span class="co">#&gt; subspot_3.1         3           1         7        17   7.33333   17.3333</span>
<span class="co">#&gt; subspot_4.1         4           1         7        18   7.33333   18.3333</span>
<span class="co">#&gt; subspot_5.1         5           1         8        13   8.33333   13.3333</span>
<span class="co">#&gt; subspot_6.1         6           1         8        14   8.33333   14.3333</span>
<span class="co">#&gt;              imagerow  imagecol spatial.cluster</span>
<span class="co">#&gt;             &lt;numeric&gt; &lt;numeric&gt;       &lt;numeric&gt;</span>
<span class="co">#&gt; subspot_1.1   7.33333   15.3333               1</span>
<span class="co">#&gt; subspot_2.1   7.33333   16.3333               2</span>
<span class="co">#&gt; subspot_3.1   7.33333   17.3333               2</span>
<span class="co">#&gt; subspot_4.1   7.33333   18.3333               2</span>
<span class="co">#&gt; subspot_5.1   8.33333   13.3333               1</span>
<span class="co">#&gt; subspot_6.1   8.33333   14.3333               1</span>
</pre></div>
<p>We can plot the enhanced cluster assignments as above.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="../reference/clusterPlot.html">clusterPlot</a></span>(<span class="kw">melanoma.enhanced</span>)
</pre></div>
<p><img src="figures/BayesSpace-enhance.plot-1.png" width="100%"></p>
</div>
<div id="enhancing-the-resolution-of-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#enhancing-the-resolution-of-gene-expression" class="anchor"></a>Enhancing the resolution of gene expression</h3>
<p>BayesSpace operates on the principal components of the gene expression matrix, and <code><a href="../reference/spatialEnhance.html">spatialEnhance()</a></code> therefore computes enhanced resolution PC vectors. Enhanced gene expression is not computed directly, and is instead imputed using a regression algorithm. For each gene, a model using the PC vectors of each spot is trained to predict the spot-level gene expression, and the fitted model is used to predict subspot expression from the subspot PCs.</p>
<p>Gene expression enhancement is implemented in the <code><a href="../reference/enhanceFeatures.html">enhanceFeatures()</a></code> function. BayesSpace predicts expression with <a href="https://xgboost.readthedocs.io/en/latest/"><code>xgboost</code></a> by default, but linear and Dirichlet regression are also available via the <code>model</code> argument. When using <code>xgboost</code>, we suggest automatically tuning the <code>nrounds</code> parameter by setting it to 0, although this comes at the cost of increased runtime (~4x slower than a pre-specified <code>nrounds</code> in practice).</p>
<p><code><a href="../reference/enhanceFeatures.html">enhanceFeatures()</a></code> can be used to impute subspot-level expression for all genes, or for a subset of genes of interest. Here, we’ll demonstrate by enhancing the expression of four marker genes: PMEL (melanoma), CD2 (T-cells), CD19 (B-cells), and COL1A1 (fibroblasts).</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PMEL"</span>, <span class="st">"CD2"</span>, <span class="st">"CD19"</span>, <span class="st">"COL1A1"</span>)
<span class="kw">melanoma.enhanced</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enhanceFeatures.html">enhanceFeatures</a></span>(<span class="kw">melanoma.enhanced</span>, <span class="kw">melanoma</span>,
                                     feature_names=<span class="kw">markers</span>,
                                     nrounds=<span class="fl">0</span>)
</pre></div>
<p>By default, log-normalized expression (<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts(sce)</a></code>) is imputed, although other assays or arbitrary feature matrices can be specified.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span>(<span class="kw">melanoma.enhanced</span>)[<span class="kw">markers</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>]
<span class="co">#&gt;        subspot_1.1 subspot_2.1 subspot_3.1 subspot_4.1 subspot_5.1</span>
<span class="co">#&gt; PMEL     1.9915980   2.3403625   2.7037618   2.3716977   2.3419490</span>
<span class="co">#&gt; CD2      0.8769051   0.4643852   0.3995831   0.4734218   0.7090108</span>
<span class="co">#&gt; CD19     0.8493783   0.6528369   0.4164957   0.5032829   0.5426237</span>
<span class="co">#&gt; COL1A1   0.3487490   2.9331527   0.5175455   1.6931008   1.3629673</span>
</pre></div>
<p>Diagnostic measures from each predictive model, such as <code>rmse</code> when using <code>xgboost</code>, are added to the <code>rowData</code> of the enhanced dataset.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu">rowData</span>(<span class="kw">melanoma.enhanced</span>)[<span class="kw">markers</span>, ]
<span class="co">#&gt; DataFrame with 4 rows and 4 columns</span>
<span class="co">#&gt;                gene_id   gene_name    is.HVG enhanceFeatures.rmse</span>
<span class="co">#&gt;            &lt;character&gt; &lt;character&gt; &lt;logical&gt;            &lt;numeric&gt;</span>
<span class="co">#&gt; PMEL   ENSG00000185664        PMEL      TRUE             0.804628</span>
<span class="co">#&gt; CD2    ENSG00000116824         CD2      TRUE             0.614575</span>
<span class="co">#&gt; CD19   ENSG00000177455        CD19      TRUE             0.697328</span>
<span class="co">#&gt; COL1A1 ENSG00000108821      COL1A1      TRUE             0.704845</span>
</pre></div>
</div>
<div id="visualizing-enhanced-gene-expression" class="section level3">
<h3 class="hasAnchor">
<a href="#visualizing-enhanced-gene-expression" class="anchor"></a>Visualizing enhanced gene expression</h3>
<p>Spatial gene expression is visualized with <code><a href="../reference/featurePlot.html">featurePlot()</a></code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="../reference/featurePlot.html">featurePlot</a></span>(<span class="kw">melanoma.enhanced</span>, <span class="st">"PMEL"</span>)
</pre></div>
<p><img src="figures/BayesSpace-enhanced.featurePlot-1.png" width="100%"></p>
<p>Here, we compare the spatial expression of the imputed marker genes.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">enhanced.plots</span> <span class="op">&lt;-</span> <span class="kw">purrr</span>::<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="kw">markers</span>, <span class="fu">function</span>(<span class="kw">x</span>) <span class="fu"><a href="../reference/featurePlot.html">featurePlot</a></span>(<span class="kw">melanoma.enhanced</span>, <span class="kw">x</span>))
<span class="kw">patchwork</span>::<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span>(<span class="kw">enhanced.plots</span>, ncol=<span class="fl">2</span>)
</pre></div>
<p><img src="figures/BayesSpace-enhanced.markers-1.png" width="100%"></p>
<p>And we can compare to the spot-level expression.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">spot.plots</span> <span class="op">&lt;-</span> <span class="kw">purrr</span>::<span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="kw">markers</span>, <span class="fu">function</span>(<span class="kw">x</span>) <span class="fu"><a href="../reference/featurePlot.html">featurePlot</a></span>(<span class="kw">melanoma</span>, <span class="kw">x</span>))
<span class="kw">patchwork</span>::<span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">enhanced.plots</span>, <span class="kw">spot.plots</span>), ncol=<span class="fl">4</span>)
</pre></div>
<p><img src="figures/BayesSpace-compare.resolution-1.png" width="100%"></p>
</div>
</div>
<div id="accessing-markov-chains" class="section level2">
<h2 class="hasAnchor">
<a href="#accessing-markov-chains" class="anchor"></a>Accessing Markov chains</h2>
<p>If <code>save.chain</code> is set to <code>TRUE</code> in either <code><a href="../reference/spatialCluster.html">spatialCluster()</a></code> or <code><a href="../reference/spatialEnhance.html">spatialEnhance()</a></code>, the chain associated with the respective MCMC run is preserved to disk as an HDF5 file. The path to this file is stored in the SingleCellExperiment’s metadata at <code>metadata(sce)$h5.chain</code>, and can be read directly using <code><a href="../reference/mcmcChain.html">mcmcChain()</a></code>.</p>
<p>The chain is provided as a <code><a href="https://rdrr.io/pkg/coda/man/mcmc.html">coda::mcmc</a></code> object, which can be analyzed with <a href="https://mjskay.github.io/tidybayes/">TidyBayes</a> or as a matrix. The object has one row per iteration, with the values of the parameters concatenated across the row. Columns are named with the parameter name and index (if any).</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">chain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcmcChain.html">mcmcChain</a></span>(<span class="kw">melanoma</span>)
<span class="kw">chain</span>[<span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>]
<span class="co">#&gt;      lambda[1,1]   lambda[1,2] lambda[1,3] lambda[1,4]  lambda[1,5]</span>
<span class="co">#&gt; [1,]  0.01000000  0.0000000000  0.00000000  0.00000000  0.000000000</span>
<span class="co">#&gt; [2,]  0.09722481 -0.0070321724 -0.04138590  0.06750923 -0.016571238</span>
<span class="co">#&gt; [3,]  0.10343843 -0.0089752921 -0.05442281  0.05637694 -0.002035676</span>
<span class="co">#&gt; [4,]  0.13778018 -0.0111105221 -0.06975277  0.03503423 -0.024012108</span>
<span class="co">#&gt; [5,]  0.12861416 -0.0009938442 -0.06240135  0.03770599 -0.026459410</span>
</pre></div>
<p>To remove the HDF5 file from disk and remove its path from the metadata, use <code><a href="../reference/mcmcChain.html">removeChain()</a></code>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Edward Zhao, Matt Stone.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
